mirrord-layer/src/file/ops.rs

```diff
@@ -249,6 +251,102 @@ pub(crate) fn close(fd: RawFd) -> Result<c_int, LayerError> {
 
     blocking_send_hook_message(HookMessage::CloseFileHook(closing_file))?;
 
-    file_channel_rx.blocking_recv()?;
-    Ok(0)
+    let CloseFileResponse { result } = file_channel_rx.blocking_recv()?;
+
+    if result == 0 {
+        // only remove the fd if everything went well in the agent
+        OPEN_FILES.lock().unwrap().remove(&fd);
+    }
+
+    Ok(result)
+}
+
+pub(crate) fn opendir(path: PathBuf) -> Result<*mut DIR, LayerError> {
+    debug!("opendir -> trying to opendir valid directory {:?}.", path);
+    let (dir_channel_tx, dir_channel_rx) = oneshot::channel::<OpenFileResponse>();
+
+    let flags = libc::O_RDONLY | libc::O_NONBLOCK | libc::O_DIRECTORY;
+
+    let opening_dir = OpenDirHook {
+        path,
+        flags,
+        dir_channel_tx,
+    };
+
+    blocking_send_hook_message(HookMessage::OpenDirHook(opening_dir))?;
+
+    let OpenFileResponse { fd } = dir_channel_rx.blocking_recv()?;
+
+    let fake_local_dir_name = CString::new(fd.to_string())?;
+
+    // Todo: check if we need extra flags like O_DIRECTORY here
+
+    let local_dir_fd = unsafe {
+        let local_dir_fd = libc::shm_open(
+            fake_local_dir_name.as_ptr(),
```

Comment:
This will cause the `*mut DIR` to be reported as not a directory. You can check for this on the sample side by looking at the file metadata (Rust has an `is_dir` method).